File: October 2022/Day 24/JAVA Code/1239. Maximum Length of a Concatenated String with Unique Characters.java
// T.C :- O(2^n)

class Solution {
    public int maxLength(List<String> arr) {
        int ans =0;
        for(int i=1;i<(1<<arr.size());i++) {
            int[] f = new int[26];
            boolean isValid=true;
            int total_size=0;
            for(int j=0;j<=arr.size() && isValid;j++) {
                if(isSet(i,j)) {
                    total_size+=arr.get(j).length();
                    for(int k=0;k<arr.get(j).length() && isValid;k++) {
                        int idx = arr.get(j).charAt(k)-'a';
                        if(f[idx]==0) {
                            f[idx]++;
                        } else {
                            isValid=false;
                        }
                    }
                }
            }
            if(isValid) ans = Math.max(ans,total_size);
        }
        return ans;
    }
    
    public static boolean isSet(int n,int idx) {
        return (n&(1<<idx))!=0;
    }
}

File: October 2022/Day 24/C++ Code/1239. Maximum Length of a Concatenated String with Unique Characters.cpp
// T.C :- O(2^n)

class Solution {
public:
    int maxLength(vector<string>& arr) {
        vector<int> fre(26,0);
        return f(arr,0,fre);
    }
    
    int f(vector<string>& arr,int idx,vector<int>& fre) {
        if(idx>=arr.size()) {
            return 0;
        }
        int ans =0;
        bool isValid = true;
        
        for(int i=0;i<arr[idx].length();i++) {
            int ix = arr[idx][i]-'a';
            if(fre[ix]>0) {
                isValid=false;
            }
            fre[ix]++;
        }
        
        if(isValid) {
            int a1 = f(arr,idx+1,fre)+arr[idx].length();
            ans = max(ans,a1);
        }
        
        for(int i=0;i<arr[idx].length();i++) {
            int ix = arr[idx][i]-'a';
            fre[ix]--;
        }
        
        int a2 = f(arr,idx+1,fre);
        ans = max(ans,a2);
        return ans;
        
    }
    
};
